internal/runner/nomad_manager.go

```diff
@@ -52,6 +55,44 @@ func (m *NomadRunnerManager) Claim(environmentID dto.EnvironmentID, duration int
 	return runner, nil
 }
 
+// checkPrewarmingPoolAlert checks if the prewarming pool contains enough idle runners as specified by the PrewarmingPoolThreshold
+// if not it starts an environment reload mechanism according to the PrewarmingPoolReloadTimeout.
+func (m *NomadRunnerManager) checkPrewarmingPoolAlert(environment ExecutionEnvironment) {
+	mutex := m.reloadingEnvironment[environment.ID()]
+	if !mutex.TryLock() {
+		// The environment is already about to be reloaded
+		return
+	}
+	defer mutex.Unlock()
+
+	prewarmingPoolThreshold := config.Config.Server.Alert.PrewarmingPoolThreshold
+	reloadTimeout := config.Config.Server.Alert.PrewarmingPoolReloadTimeout
+
+	if reloadTimeout == 0 || float64(environment.IdleRunnerCount())/float64(environment.PrewarmingPoolSize()) >= prewarmingPoolThreshold {
+		return
+	}
+
+	log.WithField(dto.KeyEnvironmentID, environment.ID()).Info("Prewarming Pool Alert. Checking again..")
```

Comment:
Even though I configured `prewarmingpoolreloadtimeout: 30`, I never get this line (and no reloading):

```log
time="2023-11-04T18:10:30.069365Z" level=debug code=204 duration="40µs" method=GET path=/api/v1/health user_agent=PostmanRuntime/7.34.0
[Sentry] 2023/11/04 18:10:30 Event dropped due to noopTransport usage.
time="2023-11-04T18:10:38.061950Z" level=debug msg="Handle Allocation Event" ClientStatus=running DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=c0caa42d-4cf8-5d48-2d4d-dddd694d531c package=nomad
time="2023-11-04T18:10:38.399774Z" level=debug msg="Handle Allocation Event" ClientStatus=complete DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=c0caa42d-4cf8-5d48-2d4d-dddd694d531c package=nomad
time="2023-11-04T18:10:38.399895Z" level=debug msg="Runner stopped" package=runner runner_id=34-0a49bd1c-7b35-11ee-836b-2e62a1fdfe46
[Sentry] 2023/11/04 18:10:51 Event dropped due to noopTransport usage.
time="2023-11-04T18:10:51.797806Z" level=error msg="Failing /api/v1/health" code=500 duration="168µs" method=GET path=/api/v1/health user_agent=PostmanRuntime/7.34.0
[Sentry] 2023/11/04 18:10:51 Event dropped due to noopTransport usage.
time="2023-11-04T18:12:01.034826Z" level=debug msg="Handle Allocation Event" ClientStatus=running DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=4b88183f-581b-3d2d-d7a1-eeafd374af08 package=nomad
time="2023-11-04T18:12:01.204157Z" level=debug msg="Handle Allocation Event" ClientStatus=complete DesiredStatus=stop NextAllocation= PrevAllocation= alloc_id=4b88183f-581b-3d2d-d7a1-eeafd374af08 package=nomad
time="2023-11-04T18:12:01.204246Z" level=debug msg="Runner stopped" package=runner runner_id=29-0a435ee0-7b35-11ee-836b-2e62a1fdfe46
[Sentry] 2023/11/04 18:12:03 Event dropped due to noopTransport usage.
time="2023-11-04T18:12:03.395046Z" level=error msg="Failing /api/v1/health" code=500 duration="71µs" method=GET path=/api/v1/health user_agent=PostmanRuntime/7.34.0
[Sentry] 2023/11/04 18:12:03 Event dropped due to noopTransport usage.
[Sentry] 2023/11/04 18:17:53 Event dropped due to noopTransport usage.
time="2023-11-04T18:17:53.293006Z" level=error msg="Failing /api/v1/health" code=500 duration="23µs" method=GET path=/api/v1/health user_agent=PostmanRuntime/7.34.0
[Sentry] 2023/11/04 18:17:53 Event dropped due to noopTransport usage.
```

In order to test the system, I was just synching an environment through CodeOcean, and was then using the Nomad UI to stop (and purge) the allocations. For the health route, this worked as expected :)

Reply:
That's the case because the check for the health route is triggered on each request to the `health` route and
the prewarming pool reloading is triggered on each request to the `provide` runner route.
We could also easily trigger the check on each request to the health route. However, this has its ups and downsides. On the one hand, currently, we would only resync those environments being used. On the other hand, when we check the prewarming pool on every health request, we could fix prewarming pool issues when they appear and not wait until someone is annoyed from receiving an error message.

Reply:
Ah, okay! I didn't consider that the `provide` route is being used there. From the documentation, I was (wrongly) assuming that there is a direct connection to the `health` route:

```yaml
    # The prewarming pool reload timeout (in seconds) defines for how long the low prewarming pool warning (above)
    # should be active before Poseidon automatically reloads the environment.
```

This sentence made that connection for me; therefore I am a little surprised about that (but get it now). For sure, having the reloading on the `provide` route makes sense (I could image a situation where Poseidon's `health` route is not regularly called -- the second mechanism should still work). Therefore, I would argue not to add it only to the `health` route. However, one could also argue that the whole alert-thing should be somewhat de-coupled from any route (the state change to warning also happens without a request, right?). Not sure, though. Let's shortly discuss the pros and cos later.